[tensorflow-common] check for invalid username

jgehw opened the pull request:
The `bazel` build tool used to build tensorflow does not support spaces in the username of the user running it. However, Windows allows spaces in the username (not only in the display name, which would be ok, but also in the native username string, which causes trouble). Thus add check to portfile, as this happens to users every couple of months, e.g. cf. #17578.

jgehw committed:
Merge pull request #1 from microsoft/master

incorporate changes from microsoft:master

jgehw committed:
Revert "incorporate changes from microsoft:master"

jgehw committed:
Merge pull request #2 from jgehw/revert-1-master

Revert "incorporate changes from microsoft:master"

jgehw committed:
Revert "Revert "incorporate changes from microsoft:master""

Commit:
Merge branch 'master' of https://github.com/microsoft/vcpkg

Commit:
Merge branch 'master' of https://github.com/microsoft/vcpkg

Commit:
Merge branch 'master' of https://github.com/microsoft/vcpkg

Commit:
Merge branch 'master' of https://github.com/microsoft/vcpkg

jgehw committed:
Merge branch 'master' of https://github.com/microsoft/vcpkg

jgehw committed:
Merge branch 'master' of https://github.com/microsoft/vcpkg

jgehw committed:
Merge branch 'master' of https://github.com/microsoft/vcpkg

jgehw committed:
Merge branch 'master' of https://github.com/microsoft/vcpkg

jgehw committed:
Merge branch 'master' of https://github.com/microsoft/vcpkg into master

jgehw committed:
Merge branch 'master' of https://github.com/microsoft/vcpkg into master

jgehw committed:
Merge branch 'master' of https://github.com/microsoft/vcpkg into master

jgehw committed:
Merge branch 'master' of https://github.com/microsoft/vcpkg into master

jgehw committed:
Merge branch 'master' of https://github.com/microsoft/vcpkg into master

jgehw committed:
Merge branch 'master' of https://github.com/microsoft/vcpkg into master

jgehw committed:
Merge branch 'master' of https://github.com/microsoft/vcpkg into master

jgehw committed:
Merge branch 'master' of https://github.com/microsoft/vcpkg into master

jgehw committed:
Merge branch 'master' of https://github.com/microsoft/vcpkg into master

jgehw committed:
Merge branch 'master' of https://github.com/microsoft/vcpkg into master

jgehw committed:
Merge branch 'master' of https://github.com/microsoft/vcpkg into master

jgehw committed:
Merge branch 'master' of https://github.com/microsoft/vcpkg into master

jgehw committed:
added check for spaces in username (unsupported by bazel)

pratikpc commented:
@jgehw would the output_user_root work well as a temporary work around if we could pass it?

CMake can be used to remove the space and if we can, then we pass output_user_root to Bazel or we can change location of Bazel output_user_root for all files for current build?

NancyLi1013 reviewed (commented):
Can you also update port-version in vcpkg.json file?

jgehw committed:
apply a work-around with custom build root instead of terminating build when username contains spaces

jgehw committed:
x-add-version

jgehw commented:
> @jgehw would the output_user_root work well as a temporary work around if we could pass it?

@pratikpc thanks for pointing me at this. I added the suggested work-around to the portfile.

ras0219 commented on ports/tensorflow-common/tensorflow-common.cmake:
```diff
@@ -219,8 +221,9 @@ foreach(BUILD_TYPE dbg rel)
+			# use --output_user_root to work-around too-long-path-names issue and username-with-spaces issue
 			vcpkg_execute_build_process(
-				COMMAND ${BASH} --noprofile --norc -c "'${BAZEL}' build --verbose_failures ${BUILD_OPTS} ${COPTS} ${CXXOPTS} ${LINKOPTS} --python_path='${PYTHON3}' --define=no_tensorflow_py_deps=true ///tensorflow:${BAZEL_LIB_NAME} ///tensorflow:install_headers"
+				COMMAND ${BASH} --noprofile --norc -c "'${BAZEL}' --output_user_root='${CURRENT_BUILDTREES_DIR}/.bzl' build --verbose_failures ${BUILD_OPTS} ${COPTS} ${CXXOPTS} ${LINKOPTS} --python_path='${PYTHON3}' --define=no_tensorflow_py_deps=true ///tensorflow:${BAZEL_LIB_NAME} ///tensorflow:install_headers"
```
Does `--output_user_root=` do what `TEST_TMPDIR` did? It is very important that vcpkg does not modify the global system state.

ras0219 replied:
Looks like yes: https://docs.bazel.build/versions/3.1.0/install-windows.html#step-4-optional-configure-output-directories

jgehw replied:
Almost: The output directory defaults to `~/.cache/bazel` on Linux, `/private/var/tmp` on macOS, and on Windows it defaults to `%HOME%` if set, else `%USERPROFILE%` if set, else the result of calling `SHGetKnownFolderPath()` with the `FOLDERID_Profile` flag set.
That was overridden by setting `$TEST_TMPDIR` because path names got too long on Windows. Removing this environment variable restores the defaults on non-Windows (there didn't seem to be a reason why this was set on all platforms). Providing the `--output_user_root=` parameter on Windows causes the output still to be placed beneath `${CURRENT_BUILDTREES_DIR}/.bzl` but with the difference that the output is placed directly beneath that folder whereas the `$TEST_TMPDIR` way caused further subfolders, among them `_bazel_${USERNAME}` which caused the trouble here.

In case using  `~/.cache/bazel` on Linux and `/private/var/tmp` on macOS should not be allowed, I could change it back to  `$TEST_TMPDIR` on these platforms.

jgehw replied:
Update: I decided to also use `${CURRENT_BUILDTREES_DIR}/.bzl` on non-Windows platforms as the tmp build output easily sums up to 100GB - so users only have to clean up one place to free disk space.

NancyLi1013 commented:
Can you run `vcpkg x-add-version --overwrite-version tensorflow-common` to update version files again? @jgehw

jgehw committed:
Add check for spaces in vcpkg path.
Keep non-Windows build output in vcpkg subfolders because it easily sums up to 100GB. So users have only one place to cleanup.

jgehw committed:
vcpkg x-add-version

jgehw committed:
fix typo

jgehw committed:
re-run x-add-version

strega-nil-ms commented:
This is nasty, but it's the correct solution. Thanks @jgehw :)

JackBoosY commented:
Hi @jgehw, recently we received some regressions about bazel that bazel has not released the control of _buildtrees\tensorflow-cc\.bzl\install\3d1356cc22901f271d47ddd780d95b50\A-server.ja_ after building tensorflow-cc.
This also means that the bazel process is not terminated after tensorflow-cc is built.

Do you have any ideas about this?
